Issue with ClassVar

Hi there,

So I encountered the following issue. I have a base class with some ClassVars:

```python
class MsgSpecSignatureModel(Struct):
    dependency_name_set: ClassVar[Set[str]]
    field_plugin_mappings: ClassVar[Dict[str, PluginMapping]]
    return_annotation: ClassVar[Any]
    signature_fields: Dict[str, SignatureField]

    # ...
```

I get the following error:

```text
self = <msgspec.inspect._Translator object at 0x115541ad0>

    def run(self):
        # First construct a decoder to validate the types are valid
        if self.protocol in (None, "msgpack"):
            from ._core import MsgpackDecoder as Decoder
        elif self.protocol == "json":
            from ._core import JSONDecoder as Decoder
        else:
            raise ValueError("protocol must be one of None, 'msgpack', 'json'")
    
>       Decoder(Tuple[self.types])
E       TypeError: Type 'typing.ClassVar[typing.Set[str]]' is not supported

```

--- a/docs/source/structs.rst
+++ b/docs/source/structs.rst
@@ -224,6 +224,37 @@ The generated ``__init__()`` for ``Subclass`` looks like:
 The field ordering rules for ``Struct`` types are identical to those for
 `dataclasses`, see the `dataclasses docs <dataclasses>`_ for more information.
 
+Class Variables
+---------------
+
+Like `dataclasses`, `msgspec.Struct` types will exclude any attribute
+annotations wrapped in `typing.ClassVar` from their fields.
+
+.. code-block:: python
+
+   >>> import msgspec
+
+   >>> from typing import ClassVar
+
+   >>> class Example(msgspec.Struct):
+   ...     x: int
+   ...     a_class_variable: ClassVar[int] = 2
+
+   >>> Example.a_class_variable
+   2
+
+   >>> Example(1)  # only `x` is counted as a field
+   Example(x=1)
+
+Note that if using `PEP 563`_ "postponed evaluation of annotations" (e.g.
+``from __future__ import annotations``) only the following spellings will work:
+
+- ``ClassVar`` or ``ClassVar[<type>]``
+- ``typing.ClassVar`` or ``typing.ClassVar[<type>]``
+
+Importing ``ClassVar`` or ``typing`` under an aliased name (e.g. ``import
+typing as typ`` or ``from typing import ClassVar as CV``) will not be properly
+detected.
 
 Type Validation
 ---------------
@@ -862,7 +893,8 @@ collected (leading to a memory leak).
 
 .. _type annotations: https://docs.python.org/3/library/typing.html
 .. _pattern matching: https://docs.python.org/3/reference/compound_stmts.html#the-match-statement
-.. _PEP 636: https://www.python.org/dev/peps/pep-0636/
+.. _PEP 636: https://peps.python.org/pep-0636/
+.. _PEP 563: https://peps.python.org/pep-0563/
 .. _dataclasses: https://docs.python.org/3/library/dataclasses.html
 .. _attrs: https://www.attrs.org/en/stable/index.html
 .. _pydantic: https://pydantic-docs.helpmanual.io/

--- a/msgspec/_core.c
+++ b/msgspec/_core.c
@@ -369,6 +369,8 @@ typedef struct {
     PyObject *typing_union;
     PyObject *typing_any;
     PyObject *typing_literal;
+    PyObject *typing_classvar;
+    PyObject *typing_generic_alias;
     PyObject *typing_annotated_alias;
     PyObject *get_type_hints;
     PyObject *get_typeddict_hints;
@@ -4990,6 +4992,17 @@ structmeta_check_namespace(PyObject *namespace) {
     return 0;
 }
 
+static PyObject *
+structmeta_get_module_ns(StructMetaInfo *info) {
+    PyObject *name = PyDict_GetItemString(info->namespace, "__module__");
+    if (name == NULL) return NULL;
+    PyObject *modules = PySys_GetObject("modules");
+    if (modules == NULL) return NULL;
+    PyObject *mod = PyDict_GetItem(modules, name);
+    if (mod == NULL) return NULL;
+    return PyObject_GetAttrString(mod, "__dict__");
+}
+
 static int
 structmeta_collect_base(StructMetaInfo *info, PyObject *base) {
     if ((PyTypeObject *)base == &StructMixinType) return 0;
@@ -5149,6 +5162,55 @@ error_mutable_struct:
 }
 
 static int
+structmeta_is_classvar(
+    StructMetaInfo *info, MsgspecState *mod, PyObject *ann, PyObject **module_ns
+) {
+    PyTypeObject *ann_type = Py_TYPE(ann);
+    if (ann_type == &PyUnicode_Type) {
+        Py_ssize_t ann_len;
+        const char *ann_buf = unicode_str_and_size(ann, &ann_len);
+        if (ann_len < 8) return 0;
+        if (memcmp(ann_buf, "ClassVar", 8) == 0) {
+            if (ann_len != 8 && ann_buf[8] != '[') return 0;
+            if (*module_ns == NULL) {
+                *module_ns = structmeta_get_module_ns(info);
+            }
+            if (*module_ns == NULL) return 0;
+            PyObject *temp = PyDict_GetItemString(*module_ns, "ClassVar");
+            return temp == mod->typing_classvar;
+        }
+        if (ann_len < 15) return 0;
+        if (memcmp(ann_buf, "typing.ClassVar", 15) == 0) {
+            if (ann_len != 15 && ann_buf[15] != '[') return 0;
+            if (*module_ns == NULL) {
+                *module_ns = structmeta_get_module_ns(info);
+            }
+            if (*module_ns == NULL) return 0;
+            PyObject *temp = PyDict_GetItemString(*module_ns, "typing");
+            if (temp == NULL) return 0;
+            temp = PyObject_GetAttrString(temp, "ClassVar");
+            int status = temp == mod->typing_classvar;
+            Py_DECREF(temp);
+            return status;
+        }
+    }
+    else {
+        if (ann == mod->typing_classvar) {
+            return 1;
+        }
+        else if ((PyObject *)ann_type == mod->typing_generic_alias) {
+            PyObject *temp = PyObject_GetAttr(ann, mod->str___origin__);
+            if (temp == NULL) return -1;
+            int status = temp == mod->typing_classvar;
+            Py_DECREF(temp);
+            return status;
+        }
+        return 0;
+    }
+    return 0;
+}
+
+static int
 structmeta_collect_fields(StructMetaInfo *info, MsgspecState *mod, bool kwonly) {
     PyObject *annotations = PyDict_GetItemString(
         info->namespace, "__annotations__"
@@ -5160,38 +5222,45 @@ structmeta_collect_fields(StructMetaInfo *info, MsgspecState *mod, bool kwonly) 
         return -1;
     }
 
-    PyObject *field;
+    PyObject *module_ns = NULL;
+    PyObject *field, *value;
     Py_ssize_t i = 0;
-    while (PyDict_Next(annotations, &i, &field, NULL)) {
+    while (PyDict_Next(annotations, &i, &field, &value)) {
         if (!PyUnicode_CheckExact(field)) {
             PyErr_SetString(
                 PyExc_TypeError, "__annotations__ keys must be strings"
             );
-            return -1;
+            goto error;
         }
 
         if (PyUnicode_Compare(field, mod->str___weakref__) == 0) {
             PyErr_SetString(
                 PyExc_TypeError, "Cannot have a struct field named '__weakref__'"
             );
-            return -1;
+            goto error;
         }
+        int status = structmeta_is_classvar(info, mod, value, &module_ns);
+        if (status == 1) continue;
+        if (status == -1) goto error;
 
         /* If the field is new, add it to slots */
         if (PyDict_GetItem(info->defaults_lk, field) == NULL) {
-            if (PyList_Append(info->slots, field) < 0) return -1;
+            if (PyList_Append(info->slots, field) < 0) goto error;
         }
 
         if (kwonly) {
-            if (PySet_Add(info->kwonly_fields, field) < 0) return -1;
+            if (PySet_Add(info->kwonly_fields, field) < 0) goto error;
         }
         else {
-            if (PySet_Discard(info->kwonly_fields, field) < 0) return -1;
+            if (PySet_Discard(info->kwonly_fields, field) < 0) goto error;
         }
 
-        if (structmeta_process_default(info, field) < 0) return -1;
+        if (structmeta_process_default(info, field) < 0) goto error;
     }
     return 0;
+error:
+    Py_XDECREF(module_ns);
+    return -1;
 }
 
 static int
@@ -17808,6 +17877,8 @@ msgspec_clear(PyObject *m)
     Py_CLEAR(st->typing_union);
     Py_CLEAR(st->typing_any);
     Py_CLEAR(st->typing_literal);
+    Py_CLEAR(st->typing_classvar);
+    Py_CLEAR(st->typing_generic_alias);
     Py_CLEAR(st->typing_annotated_alias);
     Py_CLEAR(st->get_type_hints);
     Py_CLEAR(st->get_typeddict_hints);
@@ -17890,6 +17961,8 @@ msgspec_traverse(PyObject *m, visitproc visit, void *arg)
     Py_VISIT(st->typing_union);
     Py_VISIT(st->typing_any);
     Py_VISIT(st->typing_literal);
+    Py_VISIT(st->typing_classvar);
+    Py_VISIT(st->typing_generic_alias);
     Py_VISIT(st->typing_annotated_alias);
     Py_VISIT(st->get_type_hints);
     Py_VISIT(st->get_typeddict_hints);
@@ -18091,6 +18164,8 @@ PyInit__core(void)
     SET_REF(typing_union, "Union");
     SET_REF(typing_any, "Any");
     SET_REF(typing_literal, "Literal");
+    SET_REF(typing_classvar, "ClassVar");
+    SET_REF(typing_generic_alias, "_GenericAlias");
     Py_DECREF(temp_module);
 
     temp_module = PyImport_ImportModule("msgspec._utils");

--- a/tests/test_struct.py
+++ b/tests/test_struct.py
@@ -16,6 +16,8 @@ import msgspec
 from msgspec import Struct, field, defstruct, replace
 from msgspec._core import nodefault
 
+from utils import temp_module
+
 
 @contextmanager
 def nogc():
@@ -1764,3 +1766,84 @@ class TestReplace:
         assert sys.getrefcount(x) == x_count
         assert sys.getrefcount(x2) == x2_count + 1
         del t2
+
+
+class TestClassVar:
+    def case1(self):
+        return """
+        from typing import ClassVar
+        from msgspec import Struct
+
+        class Ex(Struct):
+            a: int
+            cv1: ClassVar
+            b: int
+            cv2: ClassVar[int] = 1
+        """
+
+    def case2(self):
+        return """
+        import typing
+        from msgspec import Struct
+
+        class Ex(Struct):
+            a: int
+            cv1: typing.ClassVar
+            b: int
+            cv2: typing.ClassVar[int] = 1
+        """
+
+    def case3(self):
+        return """
+        import typing
+        from msgspec import Struct
+
+        ClassVar = typing.List
+
+        class Ex(Struct):
+            a: ClassVar
+            b: ClassVar[int]
+            cv2: typing.ClassVar[int] = 1
+        """
+
+    def case4(self):
+        return """
+        from typing import ClassVar, List
+        from msgspec import Struct
+
+        class typing:
+            ClassVar = List
+
+        class Ex(Struct):
+            a: typing.ClassVar
+            b: typing.ClassVar[int]
+            cv2: ClassVar[int] = 1
+        """
+
+    def case5(self):
+        """Annotations that start with `ClassVar`/`typing.ClassVar` but don't
+        end there aren't treated as false-positives"""
+        return """
+        from typing import ClassVar, List
+        from msgspec import Struct
+
+        ClassVariable = List
+
+        class typing:
+            ClassVariable = List
+
+        class Ex(Struct):
+            a: typing.ClassVariable
+            b: ClassVariable[int]
+            cv2: ClassVar[int] = 1
+        """
+
+    @pytest.mark.parametrize("case", [1, 2, 3, 4, 5])
+    @pytest.mark.parametrize("future_annotations", [True, False])
+    def test_classvar(self, case, future_annotations):
+        source = getattr(self, f"case{case}")()
+        if future_annotations:
+            source = "        from __future__ import annotations\n" + source
+        with temp_module(source) as mod:
+            assert mod.Ex.__struct_fields__ == ("a", "b")
+            assert mod.Ex.cv2 == 1

